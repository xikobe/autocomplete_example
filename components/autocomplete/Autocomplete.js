import React, { Component } from 'react';
import debounce from 'lodash/debounce';
import Api from 'lib/api';

// Components
import ResultsList from 'components/results-list/ResultsList';

class Autocomplete extends Component {
    constructor() {
        super();

        this.state = {
            inputValue: '',
            searchResults: [],
            focusedID: -1,
        };

        this.handleOnKeyDown = this.handleOnKeyDown.bind(this);
        this.handleOnChange = this.handleOnChange.bind(this);
        this.searchValue = debounce(this.searchValue.bind(this), 200);
    }

    handleOnChange({ target: { value } }) {
        this.setState({
            inputValue: value,
        });

        this.searchValue(value);
    }

    handleOnKeyDown(event) {
        const { focusedID, searchResults } = this.state;

        if (event.key === 'ArrowDown' && !!searchResults.length && focusedID < searchResults.length) {
            event.preventDefault();
            this.setState({
                focusedID: focusedID + 1,
            });
        }

        if (focusedID > -1) {
            event.preventDefault();

            switch (event.key) {
            case 'ArrowUp':
                this.setState({
                    focusedID: focusedID - 1,
                });
                break;

            case 'Escape':
                this.setState({
                    focusedID: -1,
                });
                break;

            case 'Enter':
            case ' ': {
                const focusedItem = document.getElementById(`item_${focusedID}`);

                this.setState({
                    inputValue: focusedItem.getAttribute('value'),
                    focusedID: -1,
                });
                break;
            }
            default:
            }
        }
    }

    searchValue(value) {
        const { searchResults } = this.state;

        this.setState({
            isLoading: true,
        });

        Api.get('/search/repositories', {
            params: {
                q: value,
            },
        }).then(({ data }) => {
            this.setState({
                searchResults: data.items,
                focusedID: -1,
                isLoading: false,
            });
        }).catch((error) => {
            this.setState({
                error,
                isLoading: false,
                searchResults: !value ? [] : searchResults,
            });
        });
    }

    render() {
        const {
            inputValue,
            searchResults,
            focusedID,
            error,
            isLoading,
        } = this.state;

        return (
            <div>
                <input
                    value={ inputValue }
                    onChange={ this.handleOnChange }
                    onKeyDown={ this.handleOnKeyDown }
                />
                {
                    isLoading ?
                        <p>Fetching awesome repos with: { inputValue }</p> :
                        <ResultsList
                            list={ searchResults }
                            focusedID={ focusedID }
                            error={ error }
                        />
                }
            </div>
        );
    }
}

export default Autocomplete;
