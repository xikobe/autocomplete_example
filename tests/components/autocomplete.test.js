import React from 'react';
import { shallow } from 'enzyme';
import Autocomplete from 'components/autocomplete/Autocomplete';

describe('Autocomplete', () => {
    describe('basic render', () => {
        it('should render correctly', () => {
            const tree = shallow(<Autocomplete />);

            expect(tree).toMatchSnapshot();
        });
    });

    describe('component handlers', () => {
        describe('onChange', () => {
            it('should call handleOnChange', () => {
                const mockEvent = {
                    target: {
                        value: 'foo',
                    },
                };

                const spy = jest.spyOn(Autocomplete.prototype, 'handleOnChange');
                const tree = shallow(<Autocomplete />);

                tree.find('input').simulate('change', mockEvent);

                expect(spy).toHaveBeenCalled();
            });

            it('should update state correctly', () => {
                const tree = shallow(<Autocomplete />);
                const mockEvent = {
                    target: {
                        value: 'foo',
                    },
                };

                tree.find('input').simulate('change', mockEvent);
                expect(tree.state().inputValue).toEqual(mockEvent.target.value);
            });
        });

        describe('onKeyDown', () => {
            let tree;
            let spy;
            const mockList = [
                { name: 'foo', id: 1 },
                { name: 'bar', id: 2 },
                { name: 'biz', id: 3 },
            ];

            beforeEach(() => {
                spy = jest.spyOn(Autocomplete.prototype, 'handleOnKeyDown');
                tree = shallow(<Autocomplete />);
            });
            it('should call handleOnKeyDown with correct event', () => {
                const mockEvent = { };

                tree.find('input').simulate('keyDown', mockEvent);

                expect(spy).toHaveBeenCalled();
            });

            describe('down arrow', () => {
                it('should not update focused id if list is empty', () => {
                    const mockEvent = {
                        key: 'ArrowDown',
                        preventDefault: jest.fn(),
                    };
                    const { focusedID } = tree.state();

                    tree.find('input').simulate('keyDown', mockEvent);

                    expect(tree.state().focusedID).toEqual(focusedID);
                });

                it('should update focused id if list is not empty', () => {
                    const mockEvent = {
                        key: 'ArrowDown',
                        preventDefault: jest.fn(),
                    };

                    tree.setState({
                        searchResults: mockList,
                    });

                    const { focusedID } = tree.state();

                    tree.find('input').simulate('keyDown', mockEvent);

                    expect(tree.state().focusedID).toEqual(focusedID + 1);
                });
            });

            describe('up arrow', () => {
                it('should not update focusedID', () => {
                    const mockEvent = {
                        key: 'ArrowUp',
                        preventDefault: jest.fn(),
                    };
                    const { focusedID } = tree.state();

                    tree.find('input').simulate('keyDown', mockEvent);

                    expect(tree.state().focusedID).toEqual(focusedID);
                });

                it('should update focusedID correctly', () => {
                    const mockEvent = {
                        key: 'ArrowUp',
                        preventDefault: jest.fn(),
                    };
                    tree.setState({
                        focusedID: 1,
                        searchResults: mockList,
                    });

                    tree.find('input').simulate('keyDown', mockEvent);

                    expect(tree.state().focusedID).toEqual(0);
                });
            });

            describe('Escape key', () => {
                it('should reset focusedID if already set', () => {
                    const mockEvent = {
                        key: 'Escape',
                        preventDefault: jest.fn(),
                    };

                    tree.setState({
                        focusedID: 1,
                    });

                    tree.find('input').simulate('keyDown', mockEvent);

                    expect(tree.state().focusedID).toEqual(-1);
                });
            });
        });
    });
});
