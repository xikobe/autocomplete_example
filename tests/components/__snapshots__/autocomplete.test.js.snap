// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete basic render should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Autocomplete />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onChange={[Function]}
          onKeyDown={[Function]}
          value=""
        />,
        <ResultsList
          focusedID={-1}
          list={Array []}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "onKeyDown": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": undefined,
          "focusedID": -1,
          "list": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />,
          <ResultsList
            focusedID={-1}
            list={Array []}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "onKeyDown": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": undefined,
            "focusedID": -1,
            "list": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
